vowpalwabbit/freegrad.cc

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#include <cmath>
+#include <string>
+#include <cfloat>
+#include "correctedMath.h"
+#include "gd.h"
+#include "shared_data.h"
+#include "io/logger.h"
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+using namespace VW::math;
+
+namespace logger = VW::io::logger;
+
+#define W    0  // current parameter
+#define Gsum    1  // sum of gradients
+#define Vsum    2  // sum of squared gradients
+#define H1   3  // maximum absolute value of features
+#define HT   4  // maximum gradient
+#define S    5  // sum of radios \sum_s |x_s|/h_s  
+#define prev_S    6  // sum of radios \sum_s |x_s|/h_s of the previous epoch
+
+struct freegrad_update_data
+{
+  struct freegrad* FG;
+  float update;
+  float predict;
+  float squared_norm_prediction; 
+  float squared_norm_grad;
+  float squared_norm_clipped_grad;
+  float sum_normalized_grad_norms;
+  float maximum_clipped_gradient_norm;
+  int num_features;
+};
+
+struct freegrad
+{
+  vw* all;  // features, finalize, l1, l2,
+  float epsilon;
+  bool restart;
+  bool project;
+  bool adaptiveradius;
+  float radius;
+  struct freegrad_update_data data;
+  size_t no_win_counter;
+  size_t early_stop_thres;
+  uint32_t freegrad_size;
+  double total_weight;
+};
+
+
+template <bool audit>
+void predict(freegrad& b, single_learner&, example& ec)
+{
+  size_t num_features_from_interactions = 0;
```

Comment:
Is it intended that the predict call invokes the regular GD path, rather than freegrad_predict? If so, consider adding comments walking through what is going on here, for future readers / contributors.